### ISSUE TITLE
Lab10 Peer Review - Matteo Pietro Pillitteri s314404

### ISSUE THREAD

**Issue by Matteo-Pietro-Pillitteri** [SalvatorePolito98/Computational_Intelligence24#5]
Hi Salvatore, after analyzing your code i have some advice for you. I hope that this review can help you.
 - I advise you to include in your works an explanatory README of your implentative choices and parameter tuning. This would help a lot in understanding your code.

 - Your results don’t look very good to me. By debugging your code I noticed some problems:
     -  First of all I added a line of code to have a cleare view of the course of the episodes:
       ` tie_percentage = (deuce / total_games) * 100 `
        This is to have a clearer view of the course of the episodes

     -  I ran 10 episodes to debug your code by entering the following prints during the training loop. This is the pseudo code after insert some debbugin prints:
     
        ```
        for episode in tqdm(range(num_episodes)):
            print(f'\n\n-----episode: {episode}-----\n\n')
            …
            while True:
                ..
                action = select_player_action(state,episode)
                print('X applies action: ', action)
                
	        ...player move...
                
                result = evaluate_state()
                print('returned result after X move: ', result)
                print_board()
        
                if someone wins:
                    break  
                
 	        else.. next_state
                
                opponent_action = select_opponent_action()
                print('O applies action: ', opponent_action)
                
 	        ...opponent move..
        
                result = evaluate_state()
                print('returned result after O move: ', result)
                print_board()
        
                if someone wins:
                    break  
                
                retrieves next state..
                print('reward: ', reward)
                initialize_Q(state)
                initialize_Q(new_state)
 	         print(‘Updating the q-table…’)
                ...update Q-table..
        ```
        With these prints you can see that thare are situation like this:
  	               
        ![O_wins](https://github.com/SalvatorePolito98/Computational_Intelligence24/assets/79902912/955e4993-72aa-4cc8-b433-2eb92a83c24e)
	 or              
        ![X_wins](https://github.com/SalvatorePolito98/Computational_Intelligence24/assets/79902912/0f58a5c8-10f3-419c-806e-69803eb65e51)

        That are situations where you get to a terminal state ( win or for X or O) but the q-table is not updated. Instead, by updating the Q value before determining the status of the terminal, you can make sure that the Q value correctly reflects the reward associated with the last state.
        Thus, if you print:
  
        `  print((1 - learning_rate) * Q_table[state][action] + learning_rate * (reward + discount_factor * max(Q_table[new_state].values()) - Q_table[state][action]))  `
  
        before to really update the q-table you can notice that you have all 0.0 at each iteration for all the episodes. So a good starting point is to check if the game is over after updating the Q-table 

   -  Once you have considered a proper update of the Q-table you could re-evaluate the parameters of your E-greedy strategy to improve performance. I can suggest, given the large number of episodes you use, that you reduce the amount of exploration_prob to 1% in order to encourage exploitation in the end. With exploitation of current knowledge the algorithm can avoid getting stuck in suboptimal local minimun.

 	      

  -  I appreciated the way you manage the game "status" and evaluate_state function. I find your approach more natural and clearer than treating a "state" as a tuple a set of cells occupied by X and a set of cells occupied by O
